docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+6. Then we see obj3, and s0_1 indicates that it should be deduped to s0 so we change obj3’s ref to point to s0_1 right away.
+7. Then we see obj4 which points to s1 and s1 indicates it’s stored in the hash table so we don’t need to dedup.
+
+Since we know the size of a string object trivially, we know which strings are on LOH.
+
+**Situations when we choose to not dedup**
+
+- If `InterlockCompareExchangePointer` fails because the ref was modified while we were finding a copy to dedup to (or insert into the hash table), we skip this ref.
+- If too many collisions exist for a hash code, we skip deduping for strings with that hash code. This avoids the DoS attack by creating too many strings for the same hash.
+
+**Alternate design points**
+
+- Should we calculate the hash codes for SOH strings as gen1 GCs promote them into gen2? 
+
+This would increase gen1 pause.
+
+- Should we dedup while creating LOH strings?
+
+This would reduce LOH string allocation speed and potentially by a lot.
+
+#### **How often to dedup**
+
+If there’s no new objects that get into gen2/LOH, there’s no need to run dedup as the duplicates would have already been eliminated (mostly) during the last dedup cycle. So we should use the new amount of objects these generations get to determine how often to dedup, ie, taking the following factors into consideration -
+
+- We know when new LOH strings are allocated (all LOH allocations are done via C++ helpers in gchelpers.cpp), we can take the accumulative size of new LOH strings till it reaches a certain percentage of LOH before starting the next dedup cycle.
+
+- We know how much is promoted into gen2 and we can also wait till a certain percentage of gen2 consists of newly promoted memory before starting the next dedup cycle.
+
+#### **Compatibility**
+
+The following scenarios become problematic or more problematic when deduping is on so we need ways to help users find them.
+
+- Mutating the string content
+
+Strings are supposed to be immutable. However in unsafe code you can change the string content after it’s created. Changing string content already asking for trouble without deduping – you could be changing the interned copy which means you are modifying someone else’s string which could cause completely unpredictable results for them. 
+
+The most common way is to use the fixed keyword:
+
+`fixed (char* p = str)`
+
+`{`
+
+  `p[0] = 'C';`
+
+`}`
+
+There are other ways such as 
+
+`((char*)(gcHandlePointingToString.AddrOfPinnedObject())) = 'c';`
+
+Or
+
+`Unsafe.AsRef (in str.GetPinnableReference()) = 'c';`
+
+- Locking on a string
+
+Locking on a string object is already discouraged due to a string can be interned. Having string dedup on can make this problematic more often if the string you called lock on is now deduped to a different string object. 
+
+- Reference equality
```

Comment:
- Among other things, our own System.Xml depends on it: https://github.com/dotnet/runtime/issues/9022#issuecomment-333235966 . We will need to figure out what to do about it. It is likely that the large workloads that benefit from string interning use System.Xml somewhere.

Reply:
Kestrel makes use of `.ReferenceEquals` for strings https://github.com/dotnet/aspnetcore/pull/9341 in header names by doing a set of string equality tests as `.ReferenceEquals` prior to falling back to the more expensive `OrdinalIgnoreCase` e.g. [HttpHeaders.Generated.cs#L7623-L7677](https://github.com/dotnet/aspnetcore/blob/602f4678c1e764e4e4c4bcdf26a5f83abe39ecdb/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs#L7623-L7677)

This means when the `Microsoft.Net.Http.Headers.HeaderNames` "constants" are used as the lookup key there is a faster path. `ReferenceEquals` returning `false` and then later returning `true` wouldn't be problematic here, rather advantageous. 

Hopefully the use of the `HeaderNames` constants is the most common choice, and string interning might pick up others? However, they were moved from regular `const string` to static properties as interning didn't necessarily seem to pick them up in cross assembly use (as they then get embedded in the calls site). String deduplication would resolve this; whereas stronger cross assembly interning would hit assembly load times? 

However, it would then hit the `ReferenceEquals` restriction, so likely would have a disadvantage if they changed back to consts with this feature, so should stay as static properties? (Properties have a downside that they cannot be used as values in attributes)

Reply:
I'll include this in the doc

Reply:
@benaadams `ReferenceEquals` can also flip from true to false. See the code example in the design document.

Reply:
>  can also flip from true to false

That's fine too for the examples I gave; as they fallback to an .Equals test. So they batch the `.ReferenceEquals` together (quick `cmp`); if none of them pass then it does the non-inline `.Equals(..., OrdinalIgnoreCase)`. Any performance hit from ReferenceEquals failing would be transitory, but it wouldn't effect correctness (in these examples)

Reply:
The locking issue can be practically mitigated by not deduping strings that have `BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX` set in the object header. This also allows `RuntimeHelpers.GetHashCode` to work as expected.

Reply:
This will reduce the chance of the app hitting the problems, but it won't make it reliable. The string deduping can kick in after the string is created, but before it is used for locks.

Reply:
I think there are two cases where deduping would affect locking on strings:

1. Deduping changes one reference to another between lock operations (like between `Enter` and `Exit` or between `Enter` on one thread and `Enter` on another thread). This would lead to an unexpected exception, or worse, to multiple threads entering the same lock.
2. Deduping merges two distinct references into one. This requires that two different locks use equal strings, which is probably not that likely. And the result is merging of those two locks, which will probably only lead to performance issues, though it's also possible it could lead to a deadlock.

Case 1 is certain to cause very serious issues, but not deduping strings that have the `SYNCBLKINDEX` bit set avoids the problem altogether.

Case 2 requires quite specific code patterns, and even then, the issues are probably not catastrophic. But not deduping strings with `SYNCBLKINDEX` does not help much in this case, since the merge can happen before the locks are entered.

So I think that not deduping strings with the `SYNCBLKINDEX` bit set is a very good idea, even if it doesn't prevent all locking issues.